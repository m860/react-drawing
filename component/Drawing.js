import * as React from "react"
import Anchor from "./Anchor";

type Props = {
    x: number,
    y: number,
    anchors?: Array
};

export default class Drawing extends React.PureComponent<Props> {
    static defaultProps = {
        anchors: []
    };

    getPosition(tag) {
        if (tag.type === "line") {
            return {
                x1: this.props.x + tag.props.x1,
                y1: this.props.y + tag.props.y1,
                x2: this.props.x + tag.props.x2,
                y2: this.props.y + tag.props.y2
            }
        }
        else if (tag.type === "circle") {
            return {
                cx: this.props.x + tag.props.cx,
                cy: this.props.y + tag.props.cy
            }
        }
        else if (tag.type === "text" || tag.type === "rect" || tag.type === "image") {
            return {
                x: this.props.x + tag.props.x,
                y: this.props.y + tag.props.y
            };
        }
        else if (tag.type === "ellipse") {
            return {
                cx: this.props.x + tag.props.cx,
                cy: this.props.y + tag.props.cy,
                rx: this.props.x + tag.props.rx,
                ry: this.props.y + tag.props.ry
            }
        }
        return {};
    }

    renderChildren() {
        if (this.props.children instanceof Array) {
            return this.props.children.map((children, index) => {
                return React.cloneElement(children, {
                    ...this.getPosition(children),
                    key: `${children.type}-${index}`
                })
            })
        }
        return React.cloneElement(this.props.children, {
            ...this.getPosition(this.props.children)
        });
    }

    renderAnchors() {
        return this.props.anchors.map((anchor, index) => {
            return <Anchor x={this.props.x + anchor.x} y={this.props.y + anchor.y} key={`anchor-${index}`}/>
        })
    }

    render() {
        return (
            <React.Fragment>
                {this.renderChildren()}
                {this.renderAnchors()}
            </React.Fragment>
        );
    }
}
