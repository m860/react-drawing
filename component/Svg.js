import * as React from "react"

export default class Svg extends React.Component {
    render() {
        return <svg {...this.props}></svg>
    }
}