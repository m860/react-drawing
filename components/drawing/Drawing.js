"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = undefined;

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

var _v = require("uuid/v1");

var _v2 = _interopRequireDefault(_v);

var _deepcopy = require("deepcopy");

var _deepcopy2 = _interopRequireDefault(_deepcopy);

var _D3Graph = require("../D3Graph");

var _deepmerge = require("deepmerge");

var _deepmerge2 = _interopRequireDefault(_deepmerge);

var _DrawingEmitter = require("./DrawingEmitter");

var _DrawingEmitter2 = _interopRequireDefault(_DrawingEmitter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Drawing = function () {
    function Drawing(option) {
        (0, _classCallCheck3.default)(this, Drawing);

        this.id = option.id ? option.id : (0, _v2.default)();
        this.attrs = option.attrs;
        this.selectedAttrs = option.selectedAttrs;
        this.text = option.text;
        this.ready = false;
        this.selected = false;
        if (option.anchors && option.anchors.length > 0) {
            var AnchorDrawing = require("./AnchorDrawing").default;
            this.anchors = option.anchors.map(function (anchorOpt) {
                return new AnchorDrawing(anchorOpt);
            });
        } else {
            this.anchors = [];
        }
        this.listeners = [];
    }

    (0, _createClass3.default)(Drawing, [{
        key: "applyAttrs",
        value: function applyAttrs() {
            var _this = this;

            var attrs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

            if (this.selection) {
                this.selection.call(function (self) {
                    for (var key in attrs) {
                        if (["x", "x1", "x2", "cx"].indexOf(key) >= 0) {
                            self.attr(key, _this.graph.toScreenX(attrs[key]));
                        } else if (["y", "y1", "y2", "cy"].indexOf(key) >= 0) {
                            self.attr(key, _this.graph.toScreenY(attrs[key]));
                        } else {
                            self.attr(key, attrs[key]);
                        }
                    }
                });
                this.selection.attr("shape-id", this.id);
                this.text && this.selection.text(this.text);
            } else {
                throw new Error("图形还没有初始化");
            }
        }
    }, {
        key: "initialize",
        value: function initialize(graph) {
            this.graph = graph;
            this.ready = true;
            //initial AnchorDrawing
            this.anchors.forEach(function (anchor) {
                return anchor.initialize(graph);
            });
        }
    }, {
        key: "render",
        value: function render() {
            var nextAttrs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

            var attrs = _deepmerge2.default.all([this.attrs, this.selected ? this.selectedAttrs : {}, nextAttrs]);
            this.applyAttrs(attrs);
            var position = this.getPosition();
            this.anchors.forEach(function (anchor) {
                var anchorOffset = anchor.getOffset();
                anchor.render({
                    cx: position.x + anchorOffset.x,
                    cy: position.y + anchorOffset.y
                });
            });
        }
    }, {
        key: "select",
        value: function select() {
            if (this.graph) {
                this.graph.doActionsAsync([new _D3Graph.SelectAction(this.id)]);
            }
        }
    }, {
        key: "remove",
        value: function remove() {
            if (this.selection) {
                this.selection.remove();
            }
            this.selection = null;
            this.listeners.forEach(function (listener) {
                return listener.remove();
            });
            this.anchors.forEach(function (anchor) {
                return anchor.remove();
            });
        }
    }, {
        key: "moveTo",
        value: function moveTo(vec) {
            throw new Error("Not implementation");
        }
    }, {
        key: "getPosition",
        value: function getPosition() {
            throw new Error("Not implementation");
        }
    }, {
        key: "toData",
        value: function toData() {
            return {
                type: this.type,
                option: {
                    id: this.id,
                    attrs: (0, _deepcopy2.default)(this.attrs),
                    text: this.text
                }
            };
        }
    }, {
        key: "findAnchor",
        value: function findAnchor(id) {
            return this.anchors.find(function (f) {
                return f.id === id;
            });
        }
    }, {
        key: "once",
        value: function once(name, callback) {
            var listener = _DrawingEmitter2.default.once(name, callback);
            this.listeners.push(listener);
        }
    }, {
        key: "addListener",
        value: function addListener(name, callback) {
            var listener = _DrawingEmitter2.default.addListener(name, callback);
            console.log("listen : " + name);
            this.listeners.push(listener);
        }
    }, {
        key: "emit",
        value: function emit(name, data) {
            console.log("emit : " + name);
            _DrawingEmitter2.default.emit(name, data);
        }
    }]);
    return Drawing;
}();

exports.default = Drawing;