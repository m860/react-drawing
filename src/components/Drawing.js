import type {DrawingOption, IAnchorDrawing, IDrawing, Point} from "./Types";
import UUID from "uuid/v1";
import deepcopy from "deepcopy"
import {SelectAction} from "./D3Graph";

export default class Drawing implements IDrawing {
    constructor(option: DrawingOption) {
        this.id = UUID();
        this.attrs = option.attrs;
        this.selectedAttrs = option.selectedAttrs;
        this.text = option.text;
        this.ready = false;
        this.selected = false;
        this.anchors = [];
    }

    applyAttrs(attrs = {}) {
        if (this.selection) {
            this.selection.call(self => {
                for (let key in attrs) {
                    if (key in ["x", "x1", "x2", "cx"]) {
                        self.attr(key, this.graph.toScreenX(attrs[key]));
                    }
                    else if (key in ["y", "y1", "y2", "cy"]) {
                        self.attr(key, this.graph.toScreenY(attrs[key]));
                    }
                    else {
                        self.attr(key, attrs[key]);
                    }
                }
            });
            this.selection.attr("shape-id", this.id);
            this.text && this.selection.text(this.text);
        }
    }

    initialize(graph) {
        this.graph = graph;
        this.ready = true;
    }

    render(nextAttrs = {}) {
        const attrs = Object.assign({}, this.attrs, this.selected ? this.selectedAttrs : {}, nextAttrs);
        this.applyAttrs(attrs);
        const position = this.getPosition();
        this.anchors.forEach((anchor: IAnchorDrawing) => anchor.render({
            cx: position.x,
            cy: position.y
        }))
    }

    select() {
        if (this.graph) {
            this.graph.doActionsAsync([
                new SelectAction(this.id)
            ])
        }
    }

    remove() {
        if (this.selection) {
            this.selection.remove();
        }
        this.selection = null;
    }

    moveTo(vec: Point) {
        throw new Error("Not implementation");
    }

    getPosition(): Point {
        throw new Error("Not implementation");
    }

    toData() {
        return {
            type: this.type,
            option: {
                id: this.id,
                attrs: deepcopy(this.attrs),
                text: this.text
            }
        }
    }
}