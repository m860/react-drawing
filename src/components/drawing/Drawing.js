import type {AnchorDrawingOption, DrawingOption, IAnchorDrawing, IDrawing, PathType, Point} from "../Types";
import UUID from "uuid/v1";
import deepcopy from "deepcopy"
import {SelectAction} from "../D3Graph";
import merge from "deepmerge"
import DrawingEmitter from "./DrawingEmitter";

export default class Drawing implements IDrawing {
    constructor(option: DrawingOption) {
        this.id = option.id ? option.id : UUID();
        this.attrs = option.attrs;
        this.selectedAttrs = option.selectedAttrs;
        this.text = option.text;
        this.ready = false;
        this.selected = false;
        if (option.anchors && option.anchors.length > 0) {
            const AnchorDrawing = require("./AnchorDrawing").default;
            this.anchors = option.anchors.map((anchorOpt: AnchorDrawingOption) => {
                return new AnchorDrawing(anchorOpt);
            });
        }
        else {
            this.anchors = [];
        }
        this.listeners = [];
        this.type = this.constructor.name;
    }

    applyAttrs(attrs = {}) {
        if (this.selection) {
            this.selection.call(self => {
                for (let key in attrs) {
                    if (["x", "x1", "x2", "cx"].indexOf(key) >= 0) {
                        self.attr(key, this.graph.toScreenX(attrs[key]));
                    }
                    else if (["y", "y1", "y2", "cy"].indexOf(key) >= 0) {
                        self.attr(key, this.graph.toScreenY(attrs[key]));
                    }
                    else {
                        self.attr(key, attrs[key]);
                    }
                }
            });
            this.selection.attr("shape-id", this.id);
            this.selection.attr("shape-type", this.type);
            this.text && this.selection.text(this.text);
        }
    }

    initialize(graph) {
        this.graph = graph;
        this.ready = true;
        //initial AnchorDrawing
        this.anchors.forEach((anchor: IAnchorDrawing) => anchor.initialize(graph));
    }

    render(nextAttrs = {}) {
        const attrs = merge.all([this.attrs || {}, this.selected ? this.selectedAttrs : {}, nextAttrs]);
        this.applyAttrs(attrs);
        const position = this.getPosition();
        this.anchors.forEach((anchor: IAnchorDrawing) => {
            const anchorOffset = anchor.getOffset();
            anchor.render({
                cx: position.x + anchorOffset.x,
                cy: position.y + anchorOffset.y
            })
        })
    }

    select() {
        if (this.graph) {
            this.graph.doActionsAsync([
                new SelectAction(this.id)
            ])
        }
    }

    remove() {
        if (this.selection) {
            this.selection.remove();
        }
        this.selection = null;
        this.listeners.forEach(listener => listener.remove());
        this.anchors.forEach(anchor => anchor.remove());
    }

    moveTo(vec: Point) {
        throw new Error("Not implementation");
    }

    getPosition(): Point {
        throw new Error("Not implementation");
    }

    toData() {
        return {
            type: this.type,
            option: {
                id: this.id,
                attrs: deepcopy(this.attrs),
                text: this.text
            }
        }
    }

    findAnchor(id) {
        return this.anchors.find(f => f.id === id);
    }

    once(name, callback) {
        const listener = DrawingEmitter.once(name, callback);
        this.listeners.push(listener);
    }

    addListener(name, callback) {
        const listener = DrawingEmitter.addListener(name, callback);
        console.log(`listen : ${name}`);
        this.listeners.push(listener);
    }

    emit(name, data) {
        console.log(`emit : ${name}`)
        DrawingEmitter.emit(name, data);
    }

    formatPath(data) {
        const arr = data.map((item: PathType) => {
            if ((item.x === null || item.x === undefined) || (item.y === null || item.y === undefined)) {
                return item.action;
            }
            return `${item.action} ${this.graph.toScreenX(item.x)} ${this.graph.toScreenY(item.y)}`
        });
        return arr.join(" ");
    }

    findDrawingByID(id) {
        if (this.id === id) {
            return this;
        }
        return null;
    }
}