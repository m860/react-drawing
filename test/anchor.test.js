import * as React from "react"
import renderer from "react-test-renderer"
import Svg from "../component/Svg";
import Drawing from "../component/Drawing";

test(`Drawing anchors`, () => {
    const component = renderer.create(
        <Svg>
            <Drawing x={100} y={100} anchors={[{x: 0, y: -50}, {x: 50, y: 0}]}>
                <rect x={-50} y={-50} width={100} height={100}></rect>
            </Drawing>
        </Svg>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})