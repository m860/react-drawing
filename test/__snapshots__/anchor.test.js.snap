// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawing anchors 1`] = `
<svg>
  <rect
    height={100}
    width={100}
    x={50}
    y={50}
  />
  <rect
    fill="black"
    height={6}
    stroke="black"
    width={6}
    x={97}
    y={47}
  />
  <rect
    fill="black"
    height={6}
    stroke="black"
    width={6}
    x={147}
    y={97}
  />
</svg>
`;
